purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/BasePurchasesIntegrationTest.kt

```diff
@@ -0,0 +1,145 @@
+package com.revenuecat.purchases
+
+import android.content.Context
+import android.preference.PreferenceManager
+import androidx.test.ext.junit.rules.activityScenarioRule
+import com.revenuecat.purchases.common.BillingAbstract
+import com.revenuecat.purchases.models.StoreTransaction
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.slot
+import org.assertj.core.api.Assertions
+import org.junit.BeforeClass
+import org.junit.Rule
+import java.net.URL
+import java.util.Date
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+import kotlin.time.Duration.Companion.seconds
+
+open class BasePurchasesIntegrationTest {
+
+    companion object {
+        @BeforeClass
+        @JvmStatic
+        fun setupClass() {
+            if (!canRunIntegrationTests()) {
+                error("You need to set required constants in Constants.kt")
+            }
+        }
+
+        private fun canRunIntegrationTests() = Constants.apiKey != "REVENUECAT_API_KEY" &&
+            Constants.googlePurchaseToken != "GOOGLE_PURCHASE_TOKEN" &&
+            Constants.productIdToPurchase != "PRODUCT_ID_TO_PURCHASE"
+    }
+
+    @get:Rule
+    var activityScenarioRule = activityScenarioRule<MainActivity>()
+
+    protected val testTimeout = 7.seconds
```

Comment:
Had to increase the timeout... Honestly this is already large, but in Android we do all requests sequentially and that's taking a long time for the new tests. Also, randomly the tests will take twice as long... We will probably want to improve the performance of the integration tests soon, but for now I wanted to get them running first. Also, the fact that we are using callbacks and not coroutines makes it necessary for us to put a timeout for the whole test :(

Reply:
iOS integration tests are also not the fastest because we need to wait on purchase expirations. I think it's okay that these tests aren't the fastest, as long as unit tests are.